For-each expression must be iterable

While debugging #279, I saw the following when running javac on the output of Specimin:
```
randoop/output/PrimitiveAndWrappedTypeVarNameCollector.java:15: error: for-each not applicable to expression type
        for (VariableDeclarator vd : n.getVariables()) {
                                                   ^
  required: array or java.lang.Iterable
  found:    GetVariablesReturnType
1 error
```

It should be straightforward to write a test case that elicits this problem.

--- a/src/main/java/org/checkerframework/specimin/JavaTypeCorrect.java
+++ b/src/main/java/org/checkerframework/specimin/JavaTypeCorrect.java
@@ -168,7 +168,7 @@ class JavaTypeCorrect {
       String binOp = null;
       String firstBinOpType = null;
 
-      // This temporary is for the for-each case.
+      // These temporaries are for the for-each case.
       String loopType = null;
       boolean lookingForLoopType = false;
 

--- a/src/main/java/org/checkerframework/specimin/JavaTypeCorrect.java
+++ b/src/main/java/org/checkerframework/specimin/JavaTypeCorrect.java
@@ -157,6 +157,7 @@ class JavaTypeCorrect {
       // We support multiline error messages of the following kinds:
       // * incompatible equality constraints
       // * bad operand types for binary operators
+      // * for-each not applicable to expression type
 
       // These are temporaries for the equality constraints case.
       String[] firstConstraints = {"equality constraints: "};
@@ -166,6 +167,10 @@ class JavaTypeCorrect {
       // These are temporaries for the binary operator case.
       String binOp = null;
       String firstBinOpType = null;
+
+      // This temporary is for the for-each case.
+      String loopType = null;
+      boolean lookingForLoopType = false;
 
       StringBuilder lines = new StringBuilder("\n");
 
@@ -184,6 +189,42 @@ class JavaTypeCorrect {
         if (line.contains("not abstract and does not override abstract method")) {
           updateClassAndUnresolvedInterface(line);
         }
+
+        // For-each logic
+        if (loopType != null) {
+          // intermediate parts of the error message, which we can skip
+          if (line.contains("^")) {
+            continue;
+          }
+          if (line.contains("required: array or java.lang.Iterable")) {
+            continue;
+          }
+          // the next line should look like: "found:    GetFoosReturnType"
+          if (!line.contains("found: ")) {
+            throw new RuntimeException(
+                "could not complete a for-each correction, because encountered "
+                    + "an unexpected line in a javac error message: "
+                    + line);
+          }
+          String typeToCorrect = line.substring(line.indexOf(':') + 1).trim();
+          changeType(typeToCorrect, loopType + "[]");
+          loopType = null;
+        }
+        if (lookingForLoopType) {
+          // line should look like: "for (Foo f : b.getFoos()) {"; we want to extract the "Foo"
+          // and put it into loopType.
+          if (loopType != null) {
+            throw new RuntimeException(
+                "loopType wasn't null when trying to set a loopType: " + loopType);
+          }
+          int startIndex = line.indexOf('(') + 1;
+          loopType = line.substring(startIndex, line.indexOf(' ', startIndex));
+          lookingForLoopType = false;
+        }
+        if (line.contains("for-each not applicable to expression type")) {
+          lookingForLoopType = true;
+        }
+
         if (line.contains("error: incompatible types")
             || line.contains("error: incomparable types")) {
           updateTypeToChange(line, filePath);

--- a/src/main/java/org/checkerframework/specimin/JavaTypeCorrect.java
+++ b/src/main/java/org/checkerframework/specimin/JavaTypeCorrect.java
@@ -209,6 +209,7 @@ class JavaTypeCorrect {
           String typeToCorrect = line.substring(line.indexOf(':') + 1).trim();
           changeType(typeToCorrect, loopType + "[]");
           loopType = null;
+          continue;
         }
         if (lookingForLoopType) {
           // line should look like: "for (Foo f : b.getFoos()) {"; we want to extract the "Foo"
@@ -220,9 +221,11 @@ class JavaTypeCorrect {
           int startIndex = line.indexOf('(') + 1;
           loopType = line.substring(startIndex, line.indexOf(' ', startIndex));
           lookingForLoopType = false;
+          continue;
         }
         if (line.contains("for-each not applicable to expression type")) {
           lookingForLoopType = true;
+          continue;
         }
 
         if (line.contains("error: incompatible types")

--- a/src/test/java/org/checkerframework/specimin/ForEachIterableTest.java
+++ b/src/test/java/org/checkerframework/specimin/ForEachIterableTest.java
@@ -1,0 +1,18 @@
+package org.checkerframework.specimin;
+
+import java.io.IOException;
+import org.junit.Test;
+
+/**
+ * This test checks that if a synthetic type is used as the iterable thing in a foreach loop,
+ * Specimin does actually make it iterable.
+ */
+public class ForEachIterableTest {
+  @Test
+  public void runTest() throws IOException {
+    SpeciminTestExecutor.runTestWithoutJarPaths(
+        "foreachiterable",
+        new String[] {"com/example/Simple.java"},
+        new String[] {"com.example.Simple#bar(Baz)"});
+  }
+}

--- a/src/test/resources/foreachiterable/expected/com/example/Simple.java
+++ b/src/test/resources/foreachiterable/expected/com/example/Simple.java
@@ -1,0 +1,11 @@
+package com.example;
+
+import org.example.Baz;
+import org.example.Foo;
+
+class Simple {
+    void bar(Baz b) {
+        for (Foo f : b.getFoos()) {
+        }
+    }
+}

--- a/src/test/resources/foreachiterable/expected/org/example/Baz.java
+++ b/src/test/resources/foreachiterable/expected/org/example/Baz.java
@@ -1,0 +1,8 @@
+package org.example;
+
+public class Baz {
+
+    public Foo[] getFoos() {
+        throw new Error();
+    }
+}

--- a/src/test/resources/foreachiterable/expected/org/example/Foo.java
+++ b/src/test/resources/foreachiterable/expected/org/example/Foo.java
@@ -1,0 +1,4 @@
+package org.example;
+
+public class Foo {
+}

--- a/src/test/resources/foreachiterable/input/com/example/Simple.java
+++ b/src/test/resources/foreachiterable/input/com/example/Simple.java
@@ -1,0 +1,13 @@
+package com.example;
+
+import org.example.Baz;
+import org.example.Foo;
+
+class Simple {
+    // Target method.
+    void bar(Baz b) {
+        for (Foo f : b.getFoos()) {
+            // do something with f
+        }
+    }
+}

